sycl/include/CL/sycl/intel/sub_group.hpp

```diff
@@ -25,69 +26,167 @@ template <typename T, access::address_space Space> class multi_ptr;
-  return __spirv_GroupFMax(__spv::Scope::Subgroup, O, x);
+// Selects 8-bit, 16-bit or 32-bit type depending on size of T. If T doesn't
+// maps to mentioned types, then void is returned
+template <typename T>
+using SelectBlockTy =
+    select_apply_cl_scalar_t<T, uint8_t, uint16_t, uint32_t, void>;
+
+template <typename T, access::address_space Space>
+using AcceptableForLoadStore =
+    bool_constant<!std::is_same<void, SelectBlockTy<T>>::value &&
+                  Space == access::address_space::global_space>;
+
+template <typename T, access::address_space Space>
+T load(const multi_ptr<T, Space> src) {
+  using Ty = SelectBlockTy<T>;
+  using GlobalPtrTy = __attribute__((opencl_global)) Ty *;
+
+  Ty Ret = __spirv_SubgroupBlockReadINTEL<Ty>(
+      reinterpret_cast<GlobalPtrTy>(src.get()));
+
+  // TODO: any better representation of reinterpret cast from Ty to T for Ret?
+  return *((T *)&Ret);
 }
 
-template <typename T, __spv::GroupOperation O>
-static typename std::enable_if<
-    !detail::is_floating_point<T>::value && std::is_integral<T>::value, T>::type
-calc(T x, intel::plus<T> op) {
-  return __spirv_GroupIAdd<T>(__spv::Scope::Subgroup, O, x);
+template <int N, typename T, access::address_space Space>
+vec<T, N> load(const multi_ptr<T, Space> src) {
+  using Ty = SelectBlockTy<T>;
+  using VecTy = Ty __attribute__((ext_vector_type(N)));
+  using GlobalPtrTy = __attribute__((opencl_global)) Ty *;
+
+  VecTy Ret = __spirv_SubgroupBlockReadINTEL<VecTy>(
+      reinterpret_cast<GlobalPtrTy>(src.get()));
+
+  // TODO: any better representation of reinterpret cast from VecTy to vec<T,
+  // N>::vector_t?
+  return *((typename vec<T, N>::vector_t *)&Ret);
 }
 
-template <typename T, __spv::GroupOperation O>
-static typename std::enable_if<detail::is_floating_point<T>::value, T>::type
-calc(T x, intel::plus<T> op) {
-  return __spirv_GroupFAdd<T>(__spv::Scope::Subgroup, O, x);
+template <typename T, access::address_space Space>
+void store(multi_ptr<T, Space> dst, const T &x) {
+  using Ty = SelectBlockTy<T>;
+  using GlobalPtrTy = __attribute__((opencl_global)) Ty *;
+
+  __spirv_SubgroupBlockWriteINTEL(reinterpret_cast<GlobalPtrTy>(dst.get()),
+  // TODO: any better representation of reinterpret cast from T to Ty for x?
+                                      *((Ty *)&x));
 }
 
+template <int N, typename T, access::address_space Space>
+void store(multi_ptr<T, Space> dst, const vec<T, N> &x) {
+  using Ty = SelectBlockTy<T>;
+  using VecTy = Ty __attribute__((ext_vector_type(N)));
+  using GlobalPtrTy = __attribute__((opencl_global)) Ty *;
+
+  __spirv_SubgroupBlockWriteINTEL(reinterpret_cast<GlobalPtrTy>(dst.get()),
+  // TODO: any better representation of reinterpret cast from T to Ty for x?
+                                      *((VecTy *)&x));
+}
+
+struct GroupOpISigned {}; struct GroupOpIUnsigned {}; struct GroupOpFP {};
+
+template <class T, typename = void> struct GroupOpTag;
+
+template <class T>
```

Comment:
For unification purposes:
```suggestion
template <typename T>
```

Reply:
Done